File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    prev = None
    # result = True
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        # self.result = True
        prev = None

        return self.inorder(root)

        # return self.result
            
    def inorder(self, root):
        if root is None:
            return True
        
        if not self.inorder(root.left):
            return False

        if self.prev is not None and self.prev.val >= root.val:
            return False
        self.prev = root
        return self.inorder(root.right)

# [50, 40, 60, 35, 45, null, 65, 30, 38, 42,46,62,68]

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        if(root == null){
            return true;
        }

        return helper(root, null, null);                
    }

    private boolean helper(TreeNode root, Integer min, Integer max){
        //base
        if(root == null){
            return true;
        }
        if(max != null && root.val >= max) return false;

        if(min != null && root.val <= min) return false;
        //logic

        // boolean case1 = ;

        // boolean case2 = ;

        return helper(root.left, min, root.val) & helper(root.right, root.val, max);
    }
}